### PR TITLE
Remove unnecessary 'static bound and cleanup

### PR DESCRIPTION
**Pull request by CryZe**
The decoder had an unnecessary 'static bound. Additionally I noticed that there's a lot of (clippy) warnings, so I cleaned up the code a bit as well. There's a few warnings left, but a few of those require breaking changes.

### CODE REVIEW

File: src/buffer.rs [hunk 1]
```diff
@@ -106,8 +104,7 @@ where
 
 #[cfg(test)]
 mod tests {
-    use crate::buffer::SamplesBuffer;
-    use crate::source::Source;
+    use crate::{buffer::SamplesBuffer, source::Source};
```
Comment (est31) [inline src/buffer.rs:107]: I don't like this change, can you keep it?

File: src/conversions/sample_rate.rs [hunk 2]
```diff
@@ -133,7 +132,7 @@ where
         }
 
         // Short circuit if there are some samples waiting.
-        if self.output_buffer.len() >= 1 {
+        if !self.output_buffer.is_empty() {
```
Comment (est31) [inline src/conversions/sample_rate.rs:135]: I don't like this change, can you keep it?

File: src/conversions/sample_rate.rs [hunk 3]
```diff
@@ -192,7 +191,7 @@ where
             debug_assert!(self.next_frame.is_empty());
 
             // draining `self.current_frame`
-            if self.current_frame.len() >= 1 {
+            if !self.current_frame.is_empty() {
```
Comment (est31) [inline src/conversions/sample_rate.rs:194]: I don't like this change, can you keep it?

File: src/decoder/flac.rs [hunk 4]
```diff
@@ -1,6 +1,9 @@
-use std::io::{Read, Seek, SeekFrom};
-use std::mem;
-use std::time::Duration;
+use std::{
```
Comment (est31) [inline src/decoder/flac.rs:1]: I don't like this change, can you keep it?

File: src/decoder/mod.rs [hunk 5]
```diff
@@ -1,12 +1,14 @@
 //! Decodes samples from an audio file.
 
-use std::error::Error;
-use std::fmt;
-use std::io::{Read, Seek};
 #[allow(unused_imports)]
 use std::io::SeekFrom;
-use std::mem;
-use std::time::Duration;
+use std::{
```
Comment (est31) [inline src/decoder/mod.rs:5]: I don't like this change, can you keep it?

File: src/decoder/mp3.rs [hunk 6]
```diff
@@ -1,5 +1,7 @@
-use std::io::{Read, Seek};
-use std::time::Duration;
+use std::{
```
Comment (est31) [inline src/decoder/mp3.rs:1]: I don't like this change, can you keep it?

File: src/decoder/vorbis.rs [hunk 7]
```diff
@@ -1,6 +1,8 @@
-use std::io::{Read, Seek, SeekFrom};
-use std::time::Duration;
-use std::vec;
+use std::{
```
Comment (est31) [inline src/decoder/vorbis.rs:1]: I don't like this change, can you keep it?

File: src/decoder/wav.rs [hunk 8]
```diff
@@ -1,5 +1,7 @@
-use std::io::{Read, Seek, SeekFrom};
-use std::time::Duration;
+use std::{
```
Comment (est31) [inline src/decoder/wav.rs:1]: I don't like this change, can you keep it?

File: src/lib.rs [hunk 9]
```diff
@@ -106,9 +104,11 @@ pub mod queue;
 pub mod source;
 pub mod static_buffer;
 
-pub use crate::conversions::Sample;
-pub use crate::decoder::Decoder;
-pub use crate::sink::Sink;
-pub use crate::source::Source;
-pub use crate::spatial_sink::SpatialSink;
-pub use crate::stream::{OutputStream, OutputStreamHandle, PlayError, StreamError};
+pub use crate::{
```
Comment (est31) [inline src/lib.rs:107]: I don't like this change, can you keep it?

File: src/sink.rs [hunk 10]
```diff
@@ -1,15 +1,14 @@
...
-use std::sync::mpsc::Receiver;
-use std::sync::Arc;
-use std::sync::Mutex;
-use std::time::Duration;
-
-use crate::queue;
-use crate::source::Done;
-use crate::Sample;
-use crate::Source;
+use std::{
```
Comment (est31) [inline src/sink.rs:2]: I don't like this change, can you keep it? Only the `std::sync::atomic::Ordering` should be put into the group of `{AtomicBool, AtomicUsize, Ordering}`.

File: src/sink.rs [hunk 11]
```diff
@@ -178,9 +177,7 @@ impl Drop for Sink {
 
 #[cfg(test)]
 mod tests {
-    use crate::buffer::SamplesBuffer;
-    use crate::sink::Sink;
-    use crate::source::Source;
+    use crate::{buffer::SamplesBuffer, sink::Sink, source::Source};
```
Comment (est31) [inline src/sink.rs:180]: I don't like this change, can you keep it?

File: src/source/amplify.rs [hunk 12]
```diff
@@ -1,18 +1,14 @@
 use std::time::Duration;
 
-use crate::Sample;
-use crate::Source;
+use crate::{Sample, Source};
```
Comment (est31) [inline src/source/amplify.rs:3]: I don't like this change, can you keep it?

File: src/source/blt.rs [hunk 13]
```diff
@@ -1,5 +1,4 @@
-use std::f32::consts::PI;
-use std::time::Duration;
+use std::{f32::consts::PI, time::Duration};
```
Comment (est31) [inline src/source/blt.rs:1]: I don't like this change, can you keep it?

File: src/source/buffered.rs [hunk 14]
```diff
@@ -1,9 +1,10 @@
-use std::cmp;
-use std::sync::{Arc, Mutex};
-use std::time::Duration;
+use std::{
```
Comment (est31) [inline src/source/buffered.rs:1]: I don't like this change, can you keep it?

File: src/source/crossfade.rs [hunk 15]
```diff
@@ -1,8 +1,9 @@
 use std::time::Duration;
 
-use crate::source::{FadeIn, Mix, TakeDuration};
-use crate::Sample;
-use crate::Source;
+use crate::{
```
Comment (est31) [inline src/source/crossfade.rs:3]: I don't like this change, can you keep it?

File: src/source/empty.rs [hunk 16]
```diff
@@ -1,12 +1,17 @@
-use crate::Sample;
-use crate::Source;
-use std::marker::PhantomData;
-use std::time::Duration;
+use crate::{Sample, Source};
+use std::{marker::PhantomData, time::Duration};
```
Comment (est31) [inline src/source/empty.rs:2]: I don't like this change, can you keep it? The sample, Source change is fine, only really about the second line here.

File: src/source/from_iter.rs [hunk 17]
```diff
@@ -140,9 +139,10 @@ where
 
 #[cfg(test)]
 mod tests {
-    use crate::buffer::SamplesBuffer;
-    use crate::source::from_iter;
-    use crate::source::Source;
+    use crate::{
+        buffer::SamplesBuffer,
+        source::{from_iter, Source},
```
Comment (est31) [inline src/source/from_iter.rs:144]: I don't like this change, can you keep it?

File: src/source/periodic.rs [hunk 18]
```diff
@@ -122,10 +121,8 @@ where
 
 #[cfg(test)]
 mod tests {
-    use crate::buffer::SamplesBuffer;
-    use crate::source::Source;
-    use std::cell::RefCell;
-    use std::time::Duration;
+    use crate::{buffer::SamplesBuffer, source::Source};
```
Comment (est31) [inline src/source/periodic.rs:124]: I don't like this change, can you keep it?

File: src/source/sine.rs [hunk 19]
```diff
@@ -1,5 +1,5 @@
 use crate::Source;
-use std::time::Duration;
+use std::{f32::consts::TAU, time::Duration};
```
Comment (est31) [inline src/source/sine.rs:2]: Can you put it onto two separate lines? Also I'd prefer usage of PI not TAU.

File: src/source/skip.rs [hunk 20]
```diff
@@ -147,15 +147,14 @@ where
...
+                .unwrap_or_else(|| Duration::from_secs(0))
         })
     }
 }
 
 #[cfg(test)]
 mod tests {
-    use crate::buffer::SamplesBuffer;
-    use crate::source::Source;
+    use crate::{buffer::SamplesBuffer, source::Source};
```
Comment (est31) [inline src/source/skip.rs:157]: I don't like this change, can you keep it?

File: src/source/uniform.rs [hunk 21]
```diff
@@ -1,14 +1,8 @@
-use std::cmp;
-use std::time::Duration;
+use std::{cmp, time::Duration};
```
Comment (est31) [inline src/source/uniform.rs:1]: I don't like this change, can you keep it? The two changes below are fine, only issues with this.

File: src/source/zero.rs [hunk 22]
```diff
@@ -1,7 +1,5 @@
-use crate::Sample;
-use crate::Source;
-use std::marker::PhantomData;
-use std::time::Duration;
+use crate::{Sample, Source};
+use std::{marker::PhantomData, time::Duration};
```
Comment (est31) [inline src/source/zero.rs:2]: I don't like this change, can you keep it? The sample, source change is fine, just issues with the second line.

File: src/static_buffer.rs [hunk 23]
```diff
@@ -10,11 +10,9 @@
 //! ```
 //!
 
-use std::slice::Iter as SliceIter;
-use std::time::Duration;
+use std::{slice::Iter as SliceIter, time::Duration};
```
Comment (est31) [inline src/static_buffer.rs:13]: I don't like this change, can you keep it?

File: src/static_buffer.rs [hunk 24]
```diff
@@ -106,8 +104,7 @@ where
 
 #[cfg(test)]
 mod tests {
-    use crate::source::Source;
-    use crate::static_buffer::StaticSamplesBuffer;
+    use crate::{source::Source, static_buffer::StaticSamplesBuffer};
```
Comment (est31) [inline src/static_buffer.rs:107]: I don't like this change, can you keep it?

**Review by est31 — CHANGES_REQUESTED** [review]
Thanks for the pr. most of it is fine but there are some changes I dont like. Especially:

* `is_empty` is longer than `.len() == 0`. I disagree that the first form is any better.
* The overeager import regrouping hurts readability a lot. I don't like it. It's ok to unify `foo::bar::Baz` and `foo::bar::Bar` into `foo::bar::{Bar, Baz}` but not ok if Bar or Baz require additional path components, nor is it ok if there are nesting levels involved. *reordering* imports is fine so that everything with std follows each other etc, but not omitting the path components.

Most other changes are obvious improvements so greatly welcome. I'll merge if the comments are addressed.

File: src/conversions/sample_rate.rs (hunk 3 above)
Comment (CryZe) [reply src/conversions/sample_rate.rs:194]: There are quite a few other .is_empty() such as literally at the top of this inline diff at 191. Do I replace them with .len() >= 1 as well? Also what about .len() > 0 maybe?

Comment (est31) [reply src/conversions/sample_rate.rs:194]: Alright let's go with your code then.

**Review by est31 — APPROVED** [review]
Thanks!

**Pull request merged**